### PR TITLE
PDF and DOCX support in Write File - Feature Improvement, close #548

### LINKED ISSUE #548: The output PDF document is damaged and cannot be opened！
When I tried to output a PDF document, I got it, but it couldn't be opened when downloaded locally，Does anyone know about this problem？？
image

### PR DESCRIPTION
**Pull request by Arkajit-Datta**
### Description

The Write File Tool can now support writing the results and discussions in a PDF or a DOCX file, with proper formatting and images embedded.

### Related Issues

#548 

### Solution and Design

Libraries called pdfkit and htmldocx are being leveraged to do the writing process.
The content initially used to be stored in a Txt file using the write tool is formatted and styled through an HTML file using an LLM.
Images generated by the agent are extracted and embedded in the same HTML file.
This HTML file is now stored in various formats like PDF & DOCX.
### Test Plan

Asking the Agent to store the results in a PDF or DOCX file.
Navigate to the workspace folder to find the generated PDF or DOCX.
### Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs update

### Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have read the contributing guide and my code conforms to the guidelines.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have added the required tests.

### CODE REVIEW

File: superagi/resource_manager/file_manager.py [hunk 1]
```diff
@@ -45,7 +56,7 @@ def write_to_s3(self, file_name, final_path):
                 s3_helper = S3Helper()
                 s3_helper.upload_file(img, path=resource.path)
 
-    def write_file(self, file_name: str, content):
+    def write_file(self, file_name: str, content, return_file_path: bool = False):
         if self.agent_id is not None:
```
Comment (luciferlinx101) [inline superagi/resource_manager/file_manager.py:60]: Dont pass return_file_path, here in write file path param

File: superagi/models/agent_execution_feed.py [hunk 2]
```diff
@@ -47,12 +47,14 @@ def get_last_tool_response(cls, session: Session, agent_execution_id: int, tool_
             AgentExecutionFeed.role == "system").order_by(AgentExecutionFeed.created_at.desc()).all()
 
         for agent_execution_feed in agent_execution_feeds:
```
[the remark this answers is not in the document]
Comment (Arkajit-Datta) [reply superagi/models/agent_execution_feed.py:49]: Reverted the changes.

File: superagi/resource_manager/file_manager.py [hunk 3]
```diff
@@ -8,11 +8,21 @@
 from superagi.models.agent import Agent
 from superagi.models.agent_execution import AgentExecution
 from superagi.types.storage_types import StorageType
+import pdfkit
+from htmldocx import HtmlToDocx
```
[the remark this answers is not in the document]
Comment (Arkajit-Datta) [reply superagi/resource_manager/file_manager.py:12]: Sure, making a folder named Exceptions and moving the Custom exceptions inside it.

File: superagi/tools/tool_response_query_manager.py [hunk 4]
```diff
@@ -11,7 +11,7 @@ def __init__(self, session: Session, agent_execution_id: int,memory:VectorStore)
```
[the remark this answers is not in the document]
Comment (Arkajit-Datta) [reply superagi/tools/tool_response_query_manager.py:11]: So, I created a functions here which was not needed after a change in dev. I removed the function while merging

File: superagi/resource_manager/file_manager.py (hunk 1 above)
Comment (Arkajit-Datta) [reply superagi/resource_manager/file_manager.py:60]: It was aiding another function, seems like obsolete now, removing it!

File: superagi/resource_manager/file_manager.py [hunk 5]
```diff
@@ -55,25 +66,19 @@ def write_file(self, file_name: str, content):
...
-                                                                      agent=Agent.get_agent_from_id(self.session,
-                                                                                                    self.agent_id),
-                                                                      agent_execution=AgentExecution
-                                                                      .get_agent_execution_from_id(self.session,
-                                                                                                   self.agent_execution_id))
-        else:
-            final_path = ResourceHelper.get_resource_path(file_name)
+
+        logger.info(f"{file_name} - File written successfully")
+
```
[the remark this answers is not in the document]
Comment (Arkajit-Datta) [reply superagi/resource_manager/file_manager.py:76]: It was previously being used in some other function, which was removed hence removing this!

File: superagi/resource_manager/file_manager.py [hunk 6]
```diff
@@ -122,3 +176,25 @@ def get_files(self):
...
+            'pdf': self.write_pdf_file,
+            'docx': self.write_docx_file, 
+            'doc': self.write_docx_file,
+            'csv': self.write_csv_file,
+            'html': self.write_txt_file
+            # NOTE: Add more file types and corresponding functions as needed, These functions should be defined 
+        }
+        
+        if file_type in file_type_handlers:
+            return file_type_handlers[file_type](file_name, file_path, content)
```
[the remark this answers is not in the document]
Comment (Arkajit-Datta) [reply superagi/resource_manager/file_manager.py:197]: Removed the return!

File: superagi/tools/file/write_file.py [hunk 7]
```diff
@@ -47,4 +51,125 @@ def _execute(self, file_name: str, content: str):
...
+        for file_name in attached_files:
+            path = self._get_file_path(file_name=file_name)
+            if not os.path.exists(path):
+                continue
+            _, file_extension = os.path.splitext(path)
+            # Check if the file extension is in the list of image extensions
+            if file_extension.lower() in image_extensions:
+                image_paths.append(path)
+        return image_paths
+
```
[the remark this answers is not in the document]
Comment (Arkajit-Datta) [reply superagi/tools/file/write_file.py:112]: That sounds fair! Removing the function

**Comment by Arkajit-Datta** [comment]
Hey @luciferlinx101 please review the PR whenever free thanx!

File: Dockerfile [hunk 8]
```diff
@@ -3,7 +3,7 @@ FROM python:3.10-slim-bullseye AS compile-image
 WORKDIR /app
```
Comment (luciferlinx101) [inline Dockerfile:3]: Add Unit Test for all methods used or changed in the existing flow ensuring maximum code coverage

**Comment by Arkajit-Datta** [comment]
Sure @luciferlinx101 I'll do these changes as soon as possible

**Comment by luciferlinx101** [comment]
Also, add required comments/documentation as maintained in the code for all methods and classes, mark the PR checklist for the same.

File: superagi/exceptions/file_errors.py [hunk 9]
```diff
@@ -0,0 +1,10 @@
+
```
[the remark this answers is not in the document]
Comment (Arkajit-Datta) [reply superagi/exceptions/file_errors.py:1]: done

**Comment by Arkajit-Datta** [comment]
@luciferlinx101 i have added the unit tests and the other review comments, Please have a look thanx

File: superagi/models/agent_execution_feed.py [hunk 10]
```diff
@@ -52,7 +52,7 @@ def get_last_tool_response(cls, session: Session, agent_execution_id: int, tool_
             if agent_execution_feed.feed.startswith("Tool"):
                 return agent_execution_feed.feed
         return ""
-
+    
     @classmethod
```
Comment (luciferlinx101) [inline superagi/models/agent_execution_feed.py:56]: revert back unrelated files too many changes to merge at a time

File: superagi/tools/tool_response_query_manager.py [hunk 11]
```diff
@@ -11,7 +11,7 @@ def __init__(self, session: Session, agent_execution_id: int,memory:VectorStore)
 
     def get_last_response(self, tool_name: str = None):
         return AgentExecutionFeed.get_last_tool_response(self.session, self.agent_execution_id, tool_name)
-    
+      
     def get_relevant_response(self, query: str,metadata:dict, top_k: int = 5):
         documents = self.memory.get_matching_text(query, metadata=metadata)
```
Comment (luciferlinx101) [inline superagi/tools/tool_response_query_manager.py:16]: Kindly revert back above change dont touch file which are not relevant to PR

File: superagi/tools/image_generation/stable_diffusion_image_gen.py [hunk 12]
```diff
@@ -57,10 +57,8 @@ def _execute(self, prompt: str, image_names: list, width: int = 512, height: int
         data = response.json()
 
         artifacts = data['artifacts']
-        base64_strings = []
-        for artifact in artifacts:
-            base64_strings.append(artifact['base64'])
-
+        base64_strings = [artifact['base64'] for artifact in artifacts]
+        image_paths=[]
```
Comment (luciferlinx101) [inline superagi/tools/image_generation/stable_diffusion_image_gen.py:61]: This can be done in seperate PR dont touch or refactor unrelated files it makes PR to big to merge

**Review by luciferlinx101 — CHANGES_REQUESTED** [review]

File: superagi/models/agent_execution_feed.py (hunk 10 above)
Comment (Arkajit-Datta) [reply superagi/models/agent_execution_feed.py]: reverted back this change

File: superagi/tools/tool_response_query_manager.py (hunk 11 above)
Comment (Arkajit-Datta) [reply superagi/tools/tool_response_query_manager.py]: reverted back this change also

File: superagi/tools/image_generation/stable_diffusion_image_gen.py (hunk 12 above)
Comment (Arkajit-Datta) [reply superagi/tools/image_generation/stable_diffusion_image_gen.py]: Sure, reverted back this change too

**Review by boundless-asura — APPROVED** [review]

**Review by luciferlinx101 — APPROVED** [review]
